codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+                    objectKeyLoop {
+                        withBlock("map.insert(", ");") {
+                            if (keyTarget.hasTrait<EnumTrait>()) {
+                                rust(
+                                    "#T::from(key.as_escaped_str())",
+                                    symbolProvider.toSymbol(keyTarget)
+                                )
+                            } else {
+                                rust("key.to_unescaped()?.into_owned()")
+                            }
+                            rust(",")
+                            deserializeMember(shape.value, forceOptional = false)
+                        }
+                    }
+                    rust("Ok(Some(map))")
+                }
+            }
+        }
+        rust("#T(tokens)?", parser)
+    }
+
+    private fun RustWriter.deserializeStruct(shape: StructureShape) {
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val symbol = symbolProvider.toSymbol(shape)
+        val nestedParser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                "Shape" to symbol,
+                *codegenScope,
+            ) {
+                startObjectOrNull {
+                    Attribute.AllowUnusedMut.render(this)
+                    rustTemplate("let mut builder = #{Shape}::builder();", *codegenScope, "Shape" to symbol)
+                    deserializeStructInner(shape)
+                    withBlock("return Ok(Some(builder.build()", "))") {
+                        if (StructureGenerator.fallibleBuilder(shape, symbolProvider)) {
+                            rustTemplate(".map_err(|_| #{Error}::custom(\"missing field\"))?", *codegenScope)
+                        }
+                    }
+                }
+            }
+        }
+        rust("#T(tokens)?", nestedParser)
+    }
+
+    private fun RustWriter.deserializeUnion(shape: UnionShape) {
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val symbol = symbolProvider.toSymbol(shape)
+        val nestedParser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                *codegenScope,
+                "Shape" to symbol
+            ) {
+                rust("let mut variant = None;")
+                rustBlock("match tokens.next().transpose()?") {
+                    rustBlockTemplate(
+                        """
+                            Some(#{Token}::ValueNull { .. }) => return Ok(None),
```

Comment:
are unions nullable? I'm not positive they are.

Reply:
I think it depends:

```
structure SomeStruct {
    unionValue: SomeUnion,
}

// Case 1: Top-level union value is null.
// Nulls shouldn't be explicit, per the spec, but we should handle
// this case gracefully and set unionValue to None in this case:

{ "unionValue": null }

// Case 2: One variant is explicitly set to null.
// This case is invalid because there has to be exactly
// one non-null variant in the union struct per the spec:

{ "unionValue": { "someVariant": null } }

// The above is basically equivalent to:

{ "unionValue": {} }
```

The match block in question is handling Case 1.